Feat/add export synonyms

Meschreiber opened the pull request.

sepehrfakour commented:
@Meschreiber this looks awesome! Great work. 😄 

A couple small initial adjustments incoming:

1) I'm not sure we should include the actual synonyms JSON in this repo. To me it seems like they're out of scope and belong in their own repo.

2) Looks like there's a typo in the path to a mock inside your integration test which is causing Travis to fail, should be a super easy fix.

3) I'm not sure the `convertCsvToJson` method https://github.com/algolia/algolia-cli/pull/25/files#diff-59986e79831306326e7821ced91cc975R50 is the most robust approach here. I was thinking we should either leverage the `csvtojson` module we're already using elsewhere, or strictly enforce a JSON api for this command, with my preference being for the former. What do you think?

Thanks for such a fantastic PR. Amazing attention to detail. You even fixed a typo in the AddRules example text!

Meschreiber commented on tests/integration/AddSynonyms.integration.test.js:
```diff
@@ -0,0 +1,64 @@
+
+const synonymsJSONPath = path.resolve(
+  process.cwd(),
+  'mocks/addSynonyms/synonym.json'
```
Oops, yup. :)

Meschreiber commented on scripts/AddSynonyms.js:
```diff
@@ -0,0 +1,95 @@
+    }
+  }
+
+  convertCsvToJson(synonymFile, filepath) {
```
I would definitely prefer (b).  About the csvtojson module: I might not fully understand how to use it, but from what I understand it converts column titles into object keys and then stores values appropriately in an array of objects.  That would not work well since synonym csvs have no headers and are just lists of n-way synonyms. They also have to be converted to a particular format to be recognized by our API (include "type":"synonym") which is why I included that short function. I would love to understand what you think some of the edge cases are so that I could address them in the function. Let me know!

sepehrfakour replied:
You make great points. I was thinking that since the `csvtojson` module has [various options](https://github.com/Keyang/node-csvtojson#parameters) that could potentially help accommodate formatting `.csv` files into the desired JSON structure, it might be a preferable way to go, but considering [the limitations you pointed out around passing in synonyms as csv](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/adding-synonyms/how-to/managing-synonyms-from-the-dashboard/#csv) it certainly might be easier and make more sense to write a simple parsing function as you've done.

What do you think about using your existing function, then also adding some context (possibly even a link to [Algolia docs](https://www.algolia.com/doc/guides/managing-results/optimize-search-results/adding-synonyms/how-to/managing-synonyms-from-the-dashboard/#csv)) in the [readme somewhere](https://github.com/algolia/algolia-cli/tree/112997bc35913230fe5b91cf36db3a45fa5cc20a#10-add-synonyms--addsynonyms) that describes the rules/limitations around importing a synonyms csv?

Meschreiber commented:
Hey @sepehrfakour, thanks for the review!  I actually think its a good idea to have the synonyms there -- eventually I was thinking of actually having commands/options that correspond to already existing synonyms file. This could definitely become not the most scalable, unless we keep it to basics (numbers/colors, etc.) Let me know what you think of this idea.

sepehrfakour commented on readme.md:
```diff
@@ -332,7 +336,54 @@ algolia exportrules -a <algoliaAppId> -k <algoliaApiKey> -n <algoliaIndexName> -
+- `<algoliaApiKey>` | Required
+- `<algoliaIndexName>` | Required
+- `<sourceFilepath>` | Required | Path to a CSV or JSON file containing an array of synonyms objects.
+- `<batchRulesParams>` | Optional | JSON object containing options passed to `batchSynonyms()` [method](https://www.algolia.com/doc/api-reference/api-methods/batch-synonyms/).
```
I think this is a typo and we wanted `batchSynonymsParams`

Meschreiber replied:
indeed, thanks!

sepehrfakour commented on scripts/AddSynonyms.js:
```diff
@@ -0,0 +1,95 @@
+    this.start = this.start.bind(this);
+    // Define validation constants
+    this.message =
+      '\nExample: $ algolia addsynonyms -a algoliaappid -k algoliaapikey -n algoliaindexname -s sourcefilepath -p batchRulesParams\n\n';
```
Typo for `batchsynonymsparams`?

Meschreiber replied:
yuss

Meschreiber commented:
@sepehrfakour -- I removed the synonyms directory  and added some clarification to the readme (and  fixed the typos! Thank you). Let me know what you think!

sepehrfakour commented on readme.md:
```diff
@@ -342,6 +342,8 @@ algolia exportrules -a <algoliaAppId> -k <algoliaApiKey> -n <algoliaIndexName> -
 
 Import a local CSV or JSON file of synonyms to an Algolia index. Some public synonym files can be downloaded from [this repository](https://github.com/algolia/synonym-dictionaries). Disclaimer: These are not intended to be all encompassing -- edits may be needed for your use case.
 
+Note that if importing a CSV file, the expected format is file with no headers and with each row of comma-separated values being a group of synonyms for each other.
```
Maybe we add a link to the relevant Algolia docs here?

eg. https://www.algolia.com/doc/guides/managing-results/optimize-search-results/adding-synonyms/how-to/managing-synonyms-from-the-dashboard/#csv

Meschreiber replied:
Good  call!

vvo commented:
Niiiice contribution!

sepehrfakour commented:
Boom. Thanks @Meschreiber !! 🔥
